File: src/jvmMain/java/resaver/pex/VDataInt.java
package resaver.pex;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.nio.ByteBuffer;

import static resaver.pex.DataType.INTEGER;

/**
 * VData that stores an integer.
 */
public class VDataInt extends VData {

    VDataInt(int val) {
        this.VALUE = val;
    }

    @Override
    public void write(@NotNull ByteBuffer output) throws IOException {
        output.put((byte) this.getType().ordinal());
        output.putInt(this.VALUE);
    }

    @Override
    public int calculateSize() {
        return 5;
    }

    @Override
    public DataType getType() {
        return INTEGER;
    }

    @Override
    public String toString() {
        return String.format("%d", this.VALUE);
    }

    public int getValue() {
        return this.VALUE;
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 83 * hash + Integer.hashCode(this.VALUE);
        return hash;
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (this == obj) {
            return true;
        } else if (obj == null) {
            return false;
        } else if (getClass() != obj.getClass()) {
            return false;
        }
        final VDataInt other = (VDataInt) obj;
        return this.VALUE == other.VALUE;
    }

    final private int VALUE;
}

File: src/jvmMain/kotlin/resaver/pex/VDataInt.kt
package resaver.pex

import PlatformByteBuffer
import java.io.IOException

/**
 * VData that stores an integer.
 */
class VDataInt internal constructor(val value: Int) : VData() {
    @Throws(IOException::class)
    override fun write(output: PlatformByteBuffer?) {
        output?.put(type.ordinal.toByte())
        output?.putInt(value)
    }

    override fun calculateSize(): Int {
        return 5
    }

    override val type: DataType
        get() = DataType.INTEGER

    override fun toString(): String {
        return String.format("%d", value)
    }

    override fun hashCode(): Int {
        var hash = 7
        hash = 83 * hash + value.hashCode()
        return hash
    }

    override fun equals(other: Any?): Boolean {
        return when {
            this === other -> {
                true
            }
            other == null -> {
                false
            }
            javaClass != other.javaClass -> {
                false
            }
            else -> {
                val other2 = other as VDataInt
                value == other2.value
            }
        }
    }

}